// @teambit/schema/dist/schema.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schemaSchema = schemaSchema;

function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));

  _graphqlTag = function () {
    return data;
  };

  return data;
}

function schemaSchema(schema) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      extend type ComponentHost {
        getSchema(id: String!): SchemaDocs
      }

      type SchemaDocs {
        exports: [String]
      }
    `,
    resolvers: {
      ComponentHost: {
        getSchema: async (host, {
          id
        }) => {
          const componentId = await host.resolveComponentId(id);
          const component = await host.get(componentId);
          const empty = {
            exports: []
          };
          if (!component) return empty;
          const docs = await schema.getSchema(component);
          if (!docs) return empty;
          return docs;
        }
      }
    }
  };
}
