// @teambit/command-bar/dist/command-bar.ui.runtime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.flat.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.array.unscopables.flat.js");

require("core-js/modules/es.promise.js");

require("core-js/modules/es.regexp.exec.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommandBarUI = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _harmony() {
  const data = require("@teambit/harmony");

  _harmony = function () {
    return data;
  };

  return data;
}

function _mousetrap() {
  const data = _interopRequireDefault(require("mousetrap"));

  _mousetrap = function () {
    return data;
  };

  return data;
}

function _ui() {
  const data = _interopRequireWildcard(require("@teambit/ui"));

  _ui = function () {
    return data;
  };

  return data;
}

function _pubsub() {
  const data = require("@teambit/pubsub");

  _pubsub = function () {
    return data;
  };

  return data;
}

function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.is-browser");

  _uiFoundationUi = function () {
    return data;
  };

  return data;
}

function _commandBar() {
  const data = require("./ui/command-bar");

  _commandBar = function () {
    return data;
  };

  return data;
}

function _commandSearcher() {
  const data = require("./ui/command-searcher");

  _commandSearcher = function () {
    return data;
  };

  return data;
}

function _commandBar2() {
  const data = require("./command-bar.aspect");

  _commandBar2 = function () {
    return data;
  };

  return data;
}

function _commandBar3() {
  const data = require("./command-bar.commands");

  _commandBar3 = function () {
    return data;
  };

  return data;
}

function _duplicateCommandError() {
  const data = require("./duplicate-command-error");

  _duplicateCommandError = function () {
    return data;
  };

  return data;
}

function _commandBarContext() {
  const data = require("./ui/command-bar-context");

  _commandBarContext = function () {
    return data;
  };

  return data;
}

function _mousetrapStub() {
  const data = require("./mousetrap-stub");

  _mousetrapStub = function () {
    return data;
  };

  return data;
}

function _keybinding() {
  const data = require("./keybinding");

  _keybinding = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const RESULT_LIMIT = 5;

/** Quick launch actions. Use the `addSearcher` slot to extend the available actions */
class CommandBarUI {
  /** Opens the command bar */

  /** Closes the command bar */

  /** Add and autocomplete provider. To support keyboard navigation, each result should have a props `active: boolean`, and `exectue: () => any` */
  addSearcher(commandSearcher) {
    this.searcherSlot.register(commandSearcher);
    return this;
  }
  /**
   * registers a command
   */


  addCommand(...originalCommands) {
    originalCommands.forEach(({
      id: commandId
    }) => {
      if (this.getCommand(commandId) !== undefined) throw new (_duplicateCommandError().DuplicateCommandError)(commandId);
    }); // commands could mutate later on, clone to ensure immutability 👌

    const commands = originalCommands.map(x => ({
      id: x.id,
      displayName: x.displayName,
      handler: x.handler,
      keybinding: x.keybinding
    }));
    this.commandSlot.register(commands);
    commands.forEach(command => {
      if (command.keybinding) {
        this.addKeybinding(command.keybinding, command.id);
      }
    });
    this.updateCommandsSearcher();
    const updaters = commands.map(command => next => {
      command.handler = next;
    });
    return updaters;
  }
  /**
   * executes command by name, if exists.
   * @param commandId
   */


  run(commandId) {
    const commandEntry = this.getCommand(commandId);
    if (!commandEntry) return undefined;
    return commandEntry.handler();
  }
  /**
   * executes a keyboard shortcut manually
   */


  getCommand(id) {
    const relevantCommands = this.commandSlot.values().map(commands => commands.find(command => command.id === id)).filter(x => !!x);
    return relevantCommands.pop();
  }

  updateCommandsSearcher() {
    const commands = this.commandSlot.values().flat();
    this.commandSearcher.update(commands);
  }

  addKeybinding(key, command) {
    this.mousetrap.bind(key, this.run.bind(this, command));
  }

  constructor(searcherSlot, commandSlot, pubSub) {
    this.searcherSlot = searcherSlot;
    this.commandSlot = commandSlot;
    (0, _defineProperty2().default)(this, "mousetrap", _uiFoundationUi().isBrowser ? new (_mousetrap().default)() : new (_mousetrapStub().MousetrapStub)());
    (0, _defineProperty2().default)(this, "commandSearcher", new (_commandSearcher().CommandSearcher)([]));
    (0, _defineProperty2().default)(this, "open", () => {
      var _this$setVisibility;

      (_this$setVisibility = this.setVisibility) === null || _this$setVisibility === void 0 ? void 0 : _this$setVisibility.call(this, true);
      return false; // aka prevent default
    });
    (0, _defineProperty2().default)(this, "close", () => {
      var _this$setVisibility2;

      (_this$setVisibility2 = this.setVisibility) === null || _this$setVisibility2 === void 0 ? void 0 : _this$setVisibility2.call(this, false);
    });
    (0, _defineProperty2().default)(this, "trigger", key => {
      this.mousetrap.trigger(key);
    });
    (0, _defineProperty2().default)(this, "search", (term, limit = RESULT_LIMIT) => {
      const searchers = this.searcherSlot.values();
      const searcher = searchers.find(x => x.test(term));
      return (searcher === null || searcher === void 0 ? void 0 : searcher.search(term, limit)) || [];
    });
    (0, _defineProperty2().default)(this, "renderContext", ({
      children
    }) => {
      return /*#__PURE__*/_react().default.createElement(_commandBarContext().CommandBarContext.Provider, {
        value: this
      }, children);
    });
    (0, _defineProperty2().default)(this, "setVisibility", void 0);
    (0, _defineProperty2().default)(this, "getCommandBar", () => {
      return /*#__PURE__*/_react().default.createElement(_commandBar().CommandBar, {
        key: "CommandBarUI",
        search: this.search,
        commander: this
      });
    });
    this.addSearcher(this.commandSearcher);
    pubSub.sub(_commandBar2().CommandBarAspect.id, e => {
      const keyboardEvent = new KeyboardEvent(e.type, e.data);
      document.dispatchEvent(keyboardEvent);
    });
  }

  static async provider([uiUi, pubsubUI], config, [searcherSlot, commandSlots]) {
    const commandBar = new CommandBarUI(searcherSlot, commandSlots, pubsubUI);
    commandBar.addCommand({
      id: _commandBar3().commandBarCommands.open,
      handler: commandBar.open,
      displayName: 'Open command bar',
      keybinding: _keybinding().openCommandBarKeybinding
    });
    uiUi.registerHudItem(commandBar.getCommandBar());
    uiUi.registerRenderHooks({
      reactContext: commandBar.renderContext
    });
    return commandBar;
  }

}

exports.CommandBarUI = CommandBarUI;
(0, _defineProperty2().default)(CommandBarUI, "dependencies", [_ui().default, _pubsub().PubsubAspect]);
(0, _defineProperty2().default)(CommandBarUI, "slots", [_harmony().Slot.withType(), _harmony().Slot.withType()]);
(0, _defineProperty2().default)(CommandBarUI, "runtime", _ui().UIRuntime);

_commandBar2().CommandBarAspect.addRuntime(CommandBarUI);
