// @teambit/sidebar/dist/ui/overview-link/overview-link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverviewLink = OverviewLink;

function _documenterUi() {
  const data = require("@teambit/documenter.ui.separator");

  _documenterUi = function () {
    return data;
  };

  return data;
}

function _evangelistElements() {
  const data = require("@teambit/evangelist.elements.icon");

  _evangelistElements = function () {
    return data;
  };

  return data;
}

function _baseUiRouting() {
  const data = require("@teambit/base-ui.routing.nav-link");

  _baseUiRouting = function () {
    return data;
  };

  return data;
}

function _classnames() {
  const data = _interopRequireDefault(require("classnames"));

  _classnames = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _overviewLinkModule() {
  const data = _interopRequireDefault(require("./overview-link.module.scss"));

  _overviewLinkModule = function () {
    return data;
  };

  return data;
}

function OverviewLink() {
  return /*#__PURE__*/_react().default.createElement("div", {
    className: _overviewLinkModule().default.overview
  }, /*#__PURE__*/_react().default.createElement(_baseUiRouting().NavLink, {
    exact: true,
    href: "/",
    activeClassName: _overviewLinkModule().default.active,
    className: (0, _classnames().default)( // hoverable, clickable, // TODO - return these somehow @oded
    _overviewLinkModule().default.overviewLink)
  }, "Components", /*#__PURE__*/_react().default.createElement(_evangelistElements().Icon, {
    of: "comps",
    className: _overviewLinkModule().default.icon
  })), /*#__PURE__*/_react().default.createElement(_documenterUi().Separator, {
    className: _overviewLinkModule().default.separator
  }));
}
