// @teambit/generator/dist/templates/workspace-generator/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.workspaceGeneratorTemplate = void 0;

function _index() {
  const data = require("./files/index");

  _index = function () {
    return data;
  };

  return data;
}

function _aspectFile() {
  const data = require("./files/aspect-file");

  _aspectFile = function () {
    return data;
  };

  return data;
}

function _docsFile() {
  const data = require("./files/docs-file");

  _docsFile = function () {
    return data;
  };

  return data;
}

function _mainRuntime() {
  const data = require("./files/main-runtime");

  _mainRuntime = function () {
    return data;
  };

  return data;
}

function _gitIgnoreTpl() {
  const data = require("./files/git-ignore-tpl");

  _gitIgnoreTpl = function () {
    return data;
  };

  return data;
}

function _readmeTpl() {
  const data = require("./files/readme-tpl");

  _readmeTpl = function () {
    return data;
  };

  return data;
}

function _indexTpl() {
  const data = require("./files/index-tpl");

  _indexTpl = function () {
    return data;
  };

  return data;
}

function _workspaceConfigTpl() {
  const data = require("./files/workspace-config-tpl");

  _workspaceConfigTpl = function () {
    return data;
  };

  return data;
}

const workspaceGeneratorTemplate = {
  name: 'workspace-generator',
  description: 'create your own workspace generator - \nDocs: https://harmony-docs.bit.dev/extending-bit/creating-a-custom-workspace-generator',
  generateFiles: context => {
    return [{
      relativePath: 'index.ts',
      content: (0, _index().indexFile)(context),
      isMain: true
    }, {
      relativePath: `${context.name}.aspect.ts`,
      content: (0, _aspectFile().aspectFile)(context)
    }, {
      relativePath: `${context.name}.main.runtime.ts`,
      content: (0, _mainRuntime().mainRuntime)(context)
    }, {
      relativePath: `${context.name}.docs.mdx`,
      content: (0, _docsFile().docsFile)()
    }, {
      relativePath: 'template/files/git-ignore.ts',
      content: (0, _gitIgnoreTpl().gitIgnoreTemplate)()
    }, {
      relativePath: 'template/files/readme-file.ts',
      content: (0, _readmeTpl().readmeTemplate)()
    }, {
      relativePath: 'template/files/workspace-config.ts',
      content: (0, _workspaceConfigTpl().workspaceConfigTemplate)()
    }, {
      relativePath: 'template/index.ts',
      content: (0, _indexTpl().indexTemplate)()
    }];
  },
  config: {
    'teambit.harmony/aspect': {},
    'teambit.envs/envs': {
      env: 'teambit.harmony/aspect'
    }
  }
};
exports.workspaceGeneratorTemplate = workspaceGeneratorTemplate;
