// @teambit/docs/dist/overview.section.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverviewSection = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _overview() {
  const data = require("./overview");

  _overview = function () {
    return data;
  };

  return data;
}

class OverviewSection {
  constructor(
  /**
   * docs ui extension.
   */
  docs) {
    this.docs = docs;
    (0, _defineProperty2().default)(this, "navigationLink", {
      href: '',
      exact: true,
      children: 'Overview'
    });
    (0, _defineProperty2().default)(this, "route", {
      path: '',
      exact: true,
      children: /*#__PURE__*/_react().default.createElement(_overview().Overview, {
        titleBadges: this.docs.titleBadgeSlot
      })
    });
    (0, _defineProperty2().default)(this, "order", 10);
  }

}

exports.OverviewSection = OverviewSection;
