// @teambit/component/dist/ui/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function () {
    return _component().Component;
  }
});
Object.defineProperty(exports, "ComponentContext", {
  enumerable: true,
  get: function () {
    return _context().ComponentContext;
  }
});
Object.defineProperty(exports, "ComponentModel", {
  enumerable: true,
  get: function () {
    return _componentModel().ComponentModel;
  }
});
Object.defineProperty(exports, "ComponentModelProps", {
  enumerable: true,
  get: function () {
    return _componentModel().ComponentModelProps;
  }
});
Object.defineProperty(exports, "ComponentProvider", {
  enumerable: true,
  get: function () {
    return _context().ComponentProvider;
  }
});
Object.defineProperty(exports, "useComponent", {
  enumerable: true,
  get: function () {
    return _useComponent().useComponent;
  }
});

function _component() {
  const data = require("./component");

  _component = function () {
    return data;
  };

  return data;
}

function _componentModel() {
  const data = require("./component-model");

  _componentModel = function () {
    return data;
  };

  return data;
}

function _context() {
  const data = require("./context");

  _context = function () {
    return data;
  };

  return data;
}

function _useComponent() {
  const data = require("./use-component");

  _useComponent = function () {
    return data;
  };

  return data;
}
