// @teambit/builder/dist/builder.graphql.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.builderSchema = builderSchema;

function _graphqlTag() {
  const data = _interopRequireDefault(require("graphql-tag"));

  _graphqlTag = function () {
    return data;
  };

  return data;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function builderSchema(builder) {
  return {
    typeDefs: (0, _graphqlTag().default)`
      type ExtensionDescriptor {
        # extension ID.
        id: String

        # icon of the extension.
        icon: String
      }

      type ExtensionArtifact {
        # descriptor of the artifact's extension
        extensionDescriptor: ExtensionDescriptor
      }

      extend type Component {
        # list of extension artifacts.
        getArtifacts(hash: String!): [ExtensionArtifact]
      }
    `,
    resolvers: {
      Component: {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        getArtifacts: async (component, {
          hash
        }) => {// const artifacts = await builder.getArtifacts(component.id, hash);
          // return artifacts.map((artifact) => artifact.toObject());
        }
      }
    }
  };
}
