// @teambit/command-bar/dist/ui/command-bar-item/command-bar-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommandBarItem = CommandBarItem;

function _extends2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

  _extends2 = function () {
    return data;
  };

  return data;
}

function _objectWithoutProperties2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

  _objectWithoutProperties2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _classnames() {
  const data = _interopRequireDefault(require("classnames"));

  _classnames = function () {
    return data;
  };

  return data;
}

function _uiFoundationUi() {
  const data = require("@teambit/ui-foundation.ui.keycap");

  _uiFoundationUi = function () {
    return data;
  };

  return data;
}

function _baseUiText() {
  const data = require("@teambit/base-ui.text.muted-text");

  _baseUiText = function () {
    return data;
  };

  return data;
}

function _commandBarItemModule() {
  const data = _interopRequireDefault(require("./command-bar-item.module.scss"));

  _commandBarItemModule = function () {
    return data;
  };

  return data;
}

function CommandBarItem(_ref) {
  let {
    entry,
    className,
    active
  } = _ref,
      rest = (0, _objectWithoutProperties2().default)(_ref, ["entry", "className", "active"]);
  const {
    handler,
    displayName: name,
    icon,
    iconAlt,
    keybinding
  } = entry;
  const firstKeybinding = Array.isArray(keybinding) ? keybinding[0] : keybinding;
  return /*#__PURE__*/_react().default.createElement("div", (0, _extends2().default)({}, rest, {
    className: (0, _classnames().default)(className, _commandBarItemModule().default.commandBarOption, active && _commandBarItemModule().default.active),
    onMouseDown: handler
  }), icon && /*#__PURE__*/_react().default.createElement("img", {
    src: icon,
    alt: iconAlt,
    className: _commandBarItemModule().default.icon
  }), /*#__PURE__*/_react().default.createElement("div", {
    className: _commandBarItemModule().default.name
  }, name), firstKeybinding && /*#__PURE__*/_react().default.createElement(_uiFoundationUi().KeySequence, {
    className: (0, _classnames().default)(_commandBarItemModule().default.commandKeys, _baseUiText().mutedText)
  }, firstKeybinding));
}
