// @teambit/graph/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Dependency", {
  enumerable: true,
  get: function () {
    return _dependency().Dependency;
  }
});
Object.defineProperty(exports, "DuplicateDependency", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().DuplicateDependency;
  }
});
Object.defineProperty(exports, "EdgeModel", {
  enumerable: true,
  get: function () {
    return _query().EdgeModel;
  }
});
Object.defineProperty(exports, "EdgeType", {
  enumerable: true,
  get: function () {
    return _edgeType().EdgeType;
  }
});
Object.defineProperty(exports, "GraphAspect", {
  enumerable: true,
  get: function () {
    return _graph().GraphAspect;
  }
});
Object.defineProperty(exports, "GraphModel", {
  enumerable: true,
  get: function () {
    return _query().GraphModel;
  }
});
Object.defineProperty(exports, "IdGraph", {
  enumerable: true,
  get: function () {
    return _objectListToGraph().IdGraph;
  }
});
Object.defineProperty(exports, "NodeModel", {
  enumerable: true,
  get: function () {
    return _query().NodeModel;
  }
});
Object.defineProperty(exports, "VersionSubgraph", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().VersionSubgraph;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function () {
    return _graph().GraphAspect;
  }
});
Object.defineProperty(exports, "objectListToGraph", {
  enumerable: true,
  get: function () {
    return _objectListToGraph().objectListToGraph;
  }
});
Object.defineProperty(exports, "useGraph", {
  enumerable: true,
  get: function () {
    return _query().useGraph;
  }
});
Object.defineProperty(exports, "useGraphQuery", {
  enumerable: true,
  get: function () {
    return _query().useGraphQuery;
  }
});

function _graph() {
  const data = require("./graph.aspect");

  _graph = function () {
    return data;
  };

  return data;
}

function _dependency() {
  const data = require("./model/dependency");

  _dependency = function () {
    return data;
  };

  return data;
}

function _duplicateDependency() {
  const data = require("./duplicate-dependency");

  _duplicateDependency = function () {
    return data;
  };

  return data;
}

function _edgeType() {
  const data = require("./edge-type");

  _edgeType = function () {
    return data;
  };

  return data;
}

function _query() {
  const data = require("./ui/query");

  _query = function () {
    return data;
  };

  return data;
}

function _objectListToGraph() {
  const data = require("./object-list-to-graph");

  _objectListToGraph = function () {
    return data;
  };

  return data;
}
