// @teambit/react-router/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function () {
    return _baseUiRouting().Link;
  }
});
Object.defineProperty(exports, "LinkAnchor", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().LinkAnchor;
  }
});
Object.defineProperty(exports, "LinkContextProvider", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().LinkContextProvider;
  }
});
Object.defineProperty(exports, "LinkProps", {
  enumerable: true,
  get: function () {
    return _baseUiRouting().LinkProps;
  }
});
Object.defineProperty(exports, "NavLink", {
  enumerable: true,
  get: function () {
    return _baseUiRouting2().NavLink;
  }
});
Object.defineProperty(exports, "NavLinkProps", {
  enumerable: true,
  get: function () {
    return _baseUiRouting2().NavLinkProps;
  }
});
exports.ReactRouter = void 0;
Object.defineProperty(exports, "ReactRouterAspect", {
  enumerable: true,
  get: function () {
    return _reactRouter().ReactRouterAspect;
  }
});
Object.defineProperty(exports, "Routing", {
  enumerable: true,
  get: function () {
    return _routingMethod().Routing;
  }
});
Object.defineProperty(exports, "RoutingProvider", {
  enumerable: true,
  get: function () {
    return _baseUiRouting3().RoutingProvider;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useLinkContext", {
  enumerable: true,
  get: function () {
    return _uiFoundationUiReactRouter().useLinkContext;
  }
});
Object.defineProperty(exports, "useLocation", {
  enumerable: true,
  get: function () {
    return _baseUiRouting3().useLocation;
  }
});
Object.defineProperty(exports, "useRouting", {
  enumerable: true,
  get: function () {
    return _baseUiRouting3().useRouting;
  }
});

function _reactRouter() {
  const data = require("./react-router.aspect");

  _reactRouter = function () {
    return data;
  };

  return data;
}

function _routingMethod() {
  const data = require("./routing-method");

  _routingMethod = function () {
    return data;
  };

  return data;
}

function _baseUiRouting() {
  const data = require("@teambit/base-ui.routing.link");

  _baseUiRouting = function () {
    return data;
  };

  return data;
}

function _baseUiRouting2() {
  const data = require("@teambit/base-ui.routing.nav-link");

  _baseUiRouting2 = function () {
    return data;
  };

  return data;
}

function _baseUiRouting3() {
  const data = require("@teambit/base-ui.routing.routing-provider");

  _baseUiRouting3 = function () {
    return data;
  };

  return data;
}

function _uiFoundationUiReactRouter() {
  const data = require("@teambit/ui-foundation.ui.react-router.link-anchor");

  _uiFoundationUiReactRouter = function () {
    return data;
  };

  return data;
}

function _ReactRouter() {
  const data = _interopRequireWildcard(require("react-router"));

  _ReactRouter = function () {
    return data;
  };

  return data;
}

Object.defineProperty(exports, "ReactRouter", {
  enumerable: true,
  get: function () {
    return _ReactRouter();
  }
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// TODO - replace with resolveFromHost / resolveFromApp when available
// this exposes the Link components installed in bit bin, so they can use the same RoutingProvider file from node_modules
var _default = _reactRouter().ReactRouterAspect;

exports.default = _default;
