// @teambit/react/dist/templates/react-workspace-app/index.js
"use strict";

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.reactWorkspaceAppTemplate = void 0;

function _workspaceCommon() {
  const data = require("../workspace-common");

  _workspaceCommon = function () {
    return data;
  };

  return data;
}

const reactWorkspaceAppTemplate = {
  name: 'react-app',
  description: 'EXPERIMENTAL. react workspace for an app',
  hidden: true,
  generateFiles: async context => {
    return (0, _workspaceCommon().generateFiles)(context);
  },
  importComponents: () => {
    return [{
      id: 'teambit.react/templates/apps/my-app',
      path: 'apps/my-app'
    }, {
      id: 'teambit.react/templates/envs/my-react',
      path: 'envs/my-react'
    }, {
      id: 'teambit.react/templates/themes/theme',
      path: 'themes/theme'
    }, {
      id: 'teambit.react/templates/styles/colors',
      path: 'styles/colors'
    }, {
      id: 'teambit.react/templates/ui/heading',
      path: 'ui/heading'
    }];
  }
};
exports.reactWorkspaceAppTemplate = reactWorkspaceAppTemplate;
