// @teambit/node/dist/node.env.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NodeEnvType = exports.NodeEnv = void 0;

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

const NodeEnvType = 'node';
exports.NodeEnvType = NodeEnvType;

class NodeEnv {
  constructor() {
    (0, _defineProperty2().default)(this, "icon", 'https://static.bit.dev/extensions-icons/nodejs.svg');
  }

  getDependencies() {
    return {
      devDependencies: {
        '@types/jest': '26.0.20',
        '@types/node': '12.20.4',
        // This is added as dev dep since our jest file transformer uses babel plugins that require this to be installed
        '@babel/runtime': '7.12.18'
      }
    };
  }

  getPreviewConfig() {
    return {
      strategyName: 'component',
      splitComponentBundle: false
    };
  }

  async __getDescriptor() {
    return {
      type: NodeEnvType
    };
  }

}

exports.NodeEnv = NodeEnv;
