// @teambit/legacy/dist/scope/models/version.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.promise.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _mergeDeepRight2() {
  const data = _interopRequireDefault(require("ramda/src/mergeDeepRight"));

  _mergeDeepRight2 = function () {
    return data;
  };

  return data;
}

function _is2() {
  const data = _interopRequireDefault(require("ramda/src/is"));

  _is2 = function () {
    return data;
  };

  return data;
}

function _flatten2() {
  const data = _interopRequireDefault(require("ramda/src/flatten"));

  _flatten2 = function () {
    return data;
  };

  return data;
}

function _isEmpty2() {
  const data = _interopRequireDefault(require("ramda/src/isEmpty"));

  _isEmpty2 = function () {
    return data;
  };

  return data;
}

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _componentVersion() {
  const data = require("@teambit/component-version");

  _componentVersion = function () {
    return data;
  };

  return data;
}

function _bitId() {
  const data = require("../../bit-id");

  _bitId = function () {
    return data;
  };

  return data;
}

function _constants() {
  const data = require("../../constants");

  _constants = function () {
    return data;
  };

  return data;
}

function _componentSchema() {
  const data = require("../../consumer/component/component-schema");

  _componentSchema = function () {
    return data;
  };

  return data;
}

function _dependencies() {
  const data = require("../../consumer/component/dependencies");

  _dependencies = function () {
    return data;
  };

  return data;
}

function _sources() {
  const data = require("../../consumer/component/sources");

  _sources = function () {
    return data;
  };

  return data;
}

function _artifactFiles() {
  const data = require("../../consumer/component/sources/artifact-files");

  _artifactFiles = function () {
    return data;
  };

  return data;
}

function _extensionData() {
  const data = require("../../consumer/config/extension-data");

  _extensionData = function () {
    return data;
  };

  return data;
}

function _logger() {
  const data = _interopRequireDefault(require("../../logger/logger"));

  _logger = function () {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("../../utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _versionInvalid() {
  const data = _interopRequireDefault(require("../exceptions/version-invalid"));

  _versionInvalid = function () {
    return data;
  };

  return data;
}

function _objects() {
  const data = require("../objects");

  _objects = function () {
    return data;
  };

  return data;
}

function _versionValidator() {
  const data = _interopRequireDefault(require("../version-validator"));

  _versionValidator = function () {
    return data;
  };

  return data;
}

/**
 * Represent a version model in the scope
 */
class Version extends _objects().BitObject {
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
  // reason for the underscore prefix is that we already have hash as a method
  // can help debugging errors when validating Version object
  constructor(props) {
    super();
    (0, _defineProperty2().default)(this, "mainFile", void 0);
    (0, _defineProperty2().default)(this, "files", void 0);
    (0, _defineProperty2().default)(this, "dists", void 0);
    (0, _defineProperty2().default)(this, "mainDistFile", void 0);
    (0, _defineProperty2().default)(this, "compiler", void 0);
    (0, _defineProperty2().default)(this, "tester", void 0);
    (0, _defineProperty2().default)(this, "log", void 0);
    (0, _defineProperty2().default)(this, "ci", void 0);
    (0, _defineProperty2().default)(this, "specsResults", void 0);
    (0, _defineProperty2().default)(this, "docs", void 0);
    (0, _defineProperty2().default)(this, "dependencies", void 0);
    (0, _defineProperty2().default)(this, "devDependencies", void 0);
    (0, _defineProperty2().default)(this, "flattenedDependencies", void 0);
    (0, _defineProperty2().default)(this, "packageDependencies", void 0);
    (0, _defineProperty2().default)(this, "devPackageDependencies", void 0);
    (0, _defineProperty2().default)(this, "peerPackageDependencies", void 0);
    (0, _defineProperty2().default)(this, "compilerPackageDependencies", void 0);
    (0, _defineProperty2().default)(this, "testerPackageDependencies", void 0);
    (0, _defineProperty2().default)(this, "bindingPrefix", void 0);
    (0, _defineProperty2().default)(this, "schema", void 0);
    (0, _defineProperty2().default)(this, "customResolvedPaths", void 0);
    (0, _defineProperty2().default)(this, "overrides", void 0);
    (0, _defineProperty2().default)(this, "packageJsonChangedProps", void 0);
    (0, _defineProperty2().default)(this, "_hash", void 0);
    (0, _defineProperty2().default)(this, "parents", void 0);
    (0, _defineProperty2().default)(this, "extensions", void 0);
    (0, _defineProperty2().default)(this, "buildStatus", void 0);
    (0, _defineProperty2().default)(this, "componentId", void 0);
    this.mainFile = props.mainFile;
    this.files = props.files;
    this.dists = props.dists;
    this.mainDistFile = props.mainDistFile;
    this.compiler = props.compiler;
    this.tester = props.tester;
    this.log = props.log;
    this.dependencies = new (_dependencies().Dependencies)(props.dependencies);
    this.devDependencies = new (_dependencies().Dependencies)(props.devDependencies);
    this.docs = props.docs;
    this.ci = props.ci || {};
    this.specsResults = props.specsResults;
    this.flattenedDependencies = props.flattenedDependencies || new (_bitId().BitIds)();
    this.packageDependencies = props.packageDependencies || {};
    this.devPackageDependencies = props.devPackageDependencies || {};
    this.peerPackageDependencies = props.peerPackageDependencies || {};
    this.compilerPackageDependencies = props.compilerPackageDependencies || {};
    this.testerPackageDependencies = props.testerPackageDependencies || {};
    this.bindingPrefix = props.bindingPrefix;
    this.schema = props.schema;
    this.customResolvedPaths = props.customResolvedPaths;
    this.overrides = props.overrides || {};
    this.packageJsonChangedProps = props.packageJsonChangedProps || {}; // @ts-ignore yes, props.hash can be undefined here, but it gets populated as soon as Version is created

    this._hash = props.hash;
    this.parents = props.parents || [];
    this.extensions = props.extensions || _extensionData().ExtensionDataList.fromArray([]);
    this.buildStatus = props.buildStatus;
    this.componentId = props.componentId;
    this.validateVersion();
  }

  validateVersion() {
    const nonEmptyFields = ['mainFile', 'files'];
    nonEmptyFields.forEach(field => {
      if (!this[field]) {
        throw new (_versionInvalid().default)(`failed creating a version object, the field "${field}" can't be empty`);
      }
    });
  }

  id() {
    const obj = this.toObject(); // @todo: remove the entire dependencies.relativePaths from the ID (it's going to be a breaking change)

    const getDependencies = deps => {
      const clonedDependencies = deps.cloneAsString(); // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!

      return clonedDependencies.map(dependency => {
        return {
          id: dependency.id,
          relativePaths: dependency.relativePaths.map(relativePath => {
            return {
              sourceRelativePath: relativePath.sourceRelativePath,
              destinationRelativePath: relativePath.destinationRelativePath
            };
          })
        };
      });
    };

    const getExtensions = extensions => {
      const sortedConfigOnly = extensions.sortById().toConfigArray();
      return sortedConfigOnly;
    };

    const filterFunction = (val, key) => {
      if (key === 'devDependencies' || key === 'devPackageDependencies' || key === 'peerPackageDependencies' || key === 'overrides' || key === 'extensions') {
        return !(0, _isEmpty2().default)(val);
      }

      return !!val;
    };

    return JSON.stringify((0, _utils().filterObject)({
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      mainFile: obj.mainFile,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      files: obj.files,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      compiler: obj.compiler,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      tester: obj.tester,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      log: obj.log,
      dependencies: getDependencies(this.dependencies),
      devDependencies: getDependencies(this.devDependencies),
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      packageDependencies: obj.packageDependencies,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      devPackageDependencies: obj.devPackageDependencies,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      peerPackageDependencies: obj.peerPackageDependencies,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      bindingPrefix: obj.bindingPrefix,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      overrides: obj.overrides,
      extensions: getExtensions(this.extensions)
    }, filterFunction));
  }

  calculateHash() {
    return new (_objects().Ref)(_objects().BitObject.makeHash(this.id()));
  }

  hash() {
    if (!this._hash) {
      throw new Error('hash is missing from a Version object');
    }

    return new (_objects().Ref)(this._hash);
  }

  get extensionDependencies() {
    return new (_dependencies().Dependencies)(this.extensions.extensionsBitIds.map(id => new (_dependencies().Dependency)(id, [])));
  }

  getAllFlattenedDependencies() {
    return _bitId().BitIds.fromArray([...this.flattenedDependencies]);
  }

  getAllDependencies() {
    return [...this.dependencies.dependencies, ...this.devDependencies.dependencies, ...this.extensionDependencies.dependencies];
  }

  get depsIdsGroupedByType() {
    return {
      dependencies: this.dependencies.getAllIds(),
      devDependencies: this.devDependencies.getAllIds(),
      extensionDependencies: this.extensions.extensionsBitIds
    };
  }

  getAllDependenciesCloned() {
    const dependencies = [...this.dependencies.getClone(), ...this.devDependencies.getClone()];
    return new (_dependencies().Dependencies)(dependencies);
  }

  getAllDependenciesIds() {
    const allDependencies = (0, _flatten2().default)(Object.values(this.depsIdsGroupedByType));
    return _bitId().BitIds.fromArray(allDependencies);
  }

  getDependenciesIdsExcludeExtensions() {
    return _bitId().BitIds.fromArray([...this.dependencies.getAllIds(), ...this.devDependencies.getAllIds()]);
  }

  updateFlattenedDependency(currentId, newId) {
    const getUpdated = flattenedDependencies => {
      const updatedIds = flattenedDependencies.map(depId => {
        if (depId.isEqual(currentId)) return newId;
        return depId;
      });
      return _bitId().BitIds.fromArray(updatedIds);
    };

    this.flattenedDependencies = getUpdated(this.flattenedDependencies);
  }

  refs() {
    return this.refsWithOptions();
  }

  refsWithOptions(includeParents = true, includeArtifacts = true) {
    const allRefs = [];

    const extractRefsFromFiles = files => {
      const refs = files ? files.map(file => file.file) : [];
      return refs;
    };

    const files = extractRefsFromFiles(this.files);
    const dists = extractRefsFromFiles(this.dists);
    allRefs.push(...files);
    allRefs.push(...dists);

    if (includeParents) {
      allRefs.push(...this.parents);
    }

    if (includeArtifacts) {
      const artifacts = (0, _artifactFiles().getRefsFromExtensions)(this.extensions);
      allRefs.push(...artifacts);
    }

    return allRefs;
  }

  refsWithoutParents() {
    const extractRefsFromFiles = files => {
      const refs = files ? files.map(file => file.file) : [];
      return refs;
    };

    const files = extractRefsFromFiles(this.files);
    const dists = extractRefsFromFiles(this.dists);
    const artifacts = (0, _artifactFiles().getRefsFromExtensions)(this.extensions);
    return [...dists, ...files, ...artifacts].filter(ref => ref);
  }

  async collectManyObjects(repo, refs) {
    return repo.loadManyRaw(refs);
  }

  toObject() {
    const _convertFileToObject = file => {
      return {
        file: file.file.toString(),
        relativePath: file.relativePath,
        name: file.name,
        test: file.test
      };
    };

    const _convertEnvToObject = env => {
      if (typeof env === 'string') {
        return env;
      } // Store the env as string in case there is no config and files (for backward compatibility)


      if (envNameOnly(env)) {
        return env.name;
      }

      const result = {
        name: env.name,
        config: env.config,
        files: []
      };
      return result;
    };

    const _removeEmptyPackagesEnvs = pkgEnv => {
      _constants().DEPENDENCIES_FIELDS.forEach(dependencyType => {
        if (pkgEnv[dependencyType] && (0, _isEmpty2().default)(pkgEnv[dependencyType])) {
          delete pkgEnv[dependencyType];
        }
      });

      return pkgEnv;
    };

    return (0, _utils().filterObject)({
      files: this.files ? this.files.map(_convertFileToObject) : null,
      mainFile: this.mainFile,
      dists: this.dists ? this.dists.map(_convertFileToObject) : null,
      mainDistFile: this.mainDistFile,
      compiler: this.compiler ? _convertEnvToObject(this.compiler) : null,
      bindingPrefix: this.bindingPrefix || _constants().DEFAULT_BINDINGS_PREFIX,
      schema: this.schema,
      tester: this.tester ? _convertEnvToObject(this.tester) : null,
      log: {
        message: this.log.message,
        date: this.log.date,
        username: this.log.username,
        email: this.log.email
      },
      ci: this.ci,
      specsResults: this.specsResults,
      docs: this.docs,
      dependencies: this.dependencies.cloneAsObject(),
      devDependencies: this.devDependencies.cloneAsObject(),
      flattenedDependencies: this.flattenedDependencies.map(dep => dep.serialize()),
      extensions: this.extensions.toModelObjects(),
      packageDependencies: this.packageDependencies,
      devPackageDependencies: this.devPackageDependencies,
      peerPackageDependencies: this.peerPackageDependencies,
      compilerPackageDependencies: _removeEmptyPackagesEnvs(this.compilerPackageDependencies),
      testerPackageDependencies: _removeEmptyPackagesEnvs(this.testerPackageDependencies),
      customResolvedPaths: this.customResolvedPaths,
      overrides: this.overrides,
      buildStatus: this.buildStatus,
      packageJsonChangedProps: this.packageJsonChangedProps,
      parents: this.parents.map(p => p.toString())
    }, val => !!val);
  }

  validateBeforePersisting(versionStr) {
    _logger().default.trace(`validating version object, hash: ${this.hash().hash}`);

    const version = Version.parse(versionStr, this._hash);
    version.validate();
  }

  toBuffer(pretty) {
    const obj = this.toObject();
    const args = (0, _utils().getStringifyArgs)(pretty);
    const str = JSON.stringify(obj, ...args);
    if (this.validateBeforePersist) this.validateBeforePersisting(str);
    return Buffer.from(str);
  }
  /**
   * used by the super class BitObject
   */


  static parse(contents, hash) {
    const contentParsed = JSON.parse(contents);
    const {
      mainFile,
      dists,
      mainDistFile,
      files,
      compiler,
      bindingPrefix,
      schema,
      tester,
      log,
      docs,
      ci,
      specsResults,
      dependencies,
      devDependencies,
      flattenedDependencies,
      flattenedDevDependencies,
      devPackageDependencies,
      peerPackageDependencies,
      compilerPackageDependencies,
      testerPackageDependencies,
      packageDependencies,
      customResolvedPaths,
      overrides,
      packageJsonChangedProps,
      extensions,
      buildStatus,
      parents
    } = contentParsed;

    const _getDependencies = (deps = []) => {
      if (deps.length && (0, _is2().default)(String, (0, _utils().first)(deps))) {
        // backward compatibility
        // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
        return deps.map(dependency => ({
          id: _bitId().BitId.parseObsolete(dependency)
        }));
      }

      const getRelativePath = relativePath => {
        if (relativePath.importSpecifiers) {
          // backward compatibility. Before the massive validation was added, an item of
          // relativePath.importSpecifiers array could be missing the mainFile property, which is
          // an invalid ImportSpecifier. (instead the mainFile it had another importSpecifiers object).
          relativePath.importSpecifiers = relativePath.importSpecifiers.filter(importSpecifier => importSpecifier.mainFile);
          if (!relativePath.importSpecifiers.length) delete relativePath.importSpecifiers;
        }

        return relativePath;
      };

      return deps.map(dependency => {
        return {
          id: _bitId().BitId.parseBackwardCompatible(dependency.id),
          relativePaths: Array.isArray(dependency.relativePaths) ? dependency.relativePaths.map(getRelativePath) : dependency.relativePaths
        };
      });
    };

    const _getFlattenedDependencies = (deps = []) => {
      return deps.map(dep => _bitId().BitId.parseBackwardCompatible(dep));
    };

    const _groupFlattenedDependencies = () => {
      // support backward compatibility. until v15, there was both flattenedDependencies and
      // flattenedDevDependencies. since then, these both were grouped to one flattenedDependencies
      const flattenedDeps = _getFlattenedDependencies(flattenedDependencies);

      const flattenedDevDeps = _getFlattenedDependencies(flattenedDevDependencies);

      return _bitId().BitIds.fromArray([...flattenedDeps, ...flattenedDevDeps]);
    };

    const parseFile = file => {
      return {
        file: _objects().Ref.from(file.file),
        relativePath: file.relativePath,
        name: file.name,
        test: file.test
      };
    };

    const _getExtensions = (exts = []) => {
      if (exts.length) {
        const newExts = exts.map(extension => {
          if (extension.extensionId) {
            const extensionId = new (_bitId().BitId)(extension.extensionId);
            const entry = new (_extensionData().ExtensionDataEntry)(undefined, extensionId, undefined, extension.config, extension.data);
            return entry;
          }

          const entry = new (_extensionData().ExtensionDataEntry)(extension.id, undefined, extension.name, extension.config, extension.data);
          return entry;
        });
        return _extensionData().ExtensionDataList.fromModelObject(newExts);
      }

      return new (_extensionData().ExtensionDataList)();
    };

    return new Version({
      mainFile,
      files: files ? files.map(parseFile) : null,
      dists: dists ? dists.map(parseFile) : null,
      mainDistFile,
      compiler: compiler ? parseEnv(compiler) : null,
      bindingPrefix: bindingPrefix || null,
      schema: schema || undefined,
      tester: tester ? parseEnv(tester) : null,
      log: {
        message: log.message,
        date: log.date,
        username: log.username,
        email: log.email
      },
      ci,
      specsResults,
      docs,
      dependencies: _getDependencies(dependencies),
      devDependencies: _getDependencies(devDependencies),
      flattenedDependencies: _groupFlattenedDependencies(),
      devPackageDependencies,
      peerPackageDependencies,
      compilerPackageDependencies,
      testerPackageDependencies,
      packageDependencies,
      customResolvedPaths,
      overrides,
      packageJsonChangedProps,
      hash,
      parents: parents ? parents.map(p => _objects().Ref.from(p)) : [],
      extensions: _getExtensions(extensions),
      buildStatus
    });
  }
  /**
   * used by raw-object.toRealObject()
   */


  static from(versionProps, hash) {
    return Version.parse(JSON.stringify(versionProps), hash);
  }
  /**
   * Create version model object from consumer component
   * @param {*} param0
   */


  static fromComponent({
    component,
    files,
    dists,
    mainDistFile
  }) {
    const parseFile = file => {
      return {
        file: file.file.hash(),
        relativePath: file.relativePath,
        name: file.name,
        test: file.test
      };
    };

    const compiler = component.compiler ? component.compiler.toModelObject() : undefined;
    const tester = component.tester ? component.tester.toModelObject() : undefined;
    const compilerDynamicPakageDependencies = component.compiler ? component.compiler.dynamicPackageDependencies : undefined;
    const testerDynamicPakageDependencies = component.tester ? component.tester.dynamicPackageDependencies : undefined;
    if (!component.log) throw new Error('Version.fromComponent - component.log is missing');
    const version = new Version({
      mainFile: component.mainFile,
      files: files.map(parseFile),
      dists: dists ? dists.map(parseFile) : undefined,
      mainDistFile,
      compiler,
      bindingPrefix: component.bindingPrefix,
      tester,
      log: component.log,
      specsResults: component.specsResults,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      docs: component.docs,
      dependencies: component.dependencies.get(),
      devDependencies: component.devDependencies.get(),
      packageDependencies: component.packageDependencies,
      devPackageDependencies: component.devPackageDependencies,
      peerPackageDependencies: component.peerPackageDependencies,
      compilerPackageDependencies: (0, _mergeDeepRight2().default)(component.compilerPackageDependencies || {}, compilerDynamicPakageDependencies || {}),
      testerPackageDependencies: (0, _mergeDeepRight2().default)(component.testerPackageDependencies || {}, testerDynamicPakageDependencies || {}),
      flattenedDependencies: component.flattenedDependencies,
      schema: component.schema,
      customResolvedPaths: component.customResolvedPaths,
      overrides: component.overrides.componentOverridesData,
      // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
      packageJsonChangedProps: component.packageJsonChangedProps,
      extensions: component.extensions,
      buildStatus: component.buildStatus,
      componentId: component.id
    });

    if ((0, _componentVersion().isHash)(component.version)) {
      version._hash = component.version;
    } else {
      version.setNewHash();
    }

    return version;
  }

  setNewHash() {
    // @todo: after v15 is deployed, this can be changed to generate a random uuid
    this._hash = this.calculateHash().toString();
  }

  get ignoreSharedDir() {
    return !(0, _componentSchema().isSchemaSupport)(_componentSchema().SchemaFeature.sharedDir, this.schema);
  }

  get isLegacy() {
    return !this.schema || this.schema === _componentSchema().SchemaName.Legacy;
  }

  setSpecsResults(specsResults) {
    this.specsResults = specsResults;
  }

  setDist(dist) {
    // @ts-ignore AUTO-ADDED-AFTER-MIGRATION-PLEASE-FIX!
    this.dist = dist ? {
      file: dist.hash(),
      name: _constants().DEFAULT_BUNDLE_FILENAME
    } : null;
  }

  setCIProps(ci) {
    this.ci = ci;
  }

  hasParent(ref) {
    return this.parents.find(p => p.toString() === ref.toString());
  }

  addParent(ref) {
    if (this.isLegacy) return;

    if (this.hasParent(ref)) {
      return; // make sure to not add twice
    }

    this.parents.push(ref);
  }

  addAsOnlyParent(ref) {
    if (this.isLegacy) return;
    this.parents = [ref];
  }

  removeParent(ref) {
    this.parents = this.parents.filter(p => p.toString() !== ref.toString());
  }

  modelFilesToSourceFiles(repository) {
    return Promise.all(this.files.map(file => _sources().SourceFile.loadFromSourceFileModel(file, repository)));
  }
  /**
   * Validate the version model properties, to make sure we are not inserting something
   * in the wrong format
   */


  validate() {
    (0, _versionValidator().default)(this);
  }

}

exports.default = Version;

function parseEnv(env) {
  if (typeof env === 'string') {
    return env;
  } // Store the env as string in case there is no config and files (for backward compatibility)


  if (envNameOnly(env)) {
    return env.name;
  }

  return {
    name: env.name,
    config: env.config
  };
}

function envNameOnly(env) {
  if ((!env.config || (0, _isEmpty2().default)(env.config)) && (!env.files || (0, _isEmpty2().default)(env.files))) {
    return true;
  }

  return false;
}
